src/software/ai/evaluation/calc_best_shot.cpp

```diff
@@ -147,3 +151,44 @@ std::optional<Shot> calcBestShotOnGoal(const Field &field, const Team &friendly_
             obstacles, goal, radius);
     }
 }
+
+std::optional<Shot> findBestShotOnGoal(const Field &field, const Team &friendly_team,
+                                       const Team &enemy_team,
+                                       const Point &starting_point, TeamType goal,
+                                       const std::vector<Robot> &robots_to_ignore,
+                                       double radius)
+{
+    // Number of shot origin points to sample to the left and right of the starting point
+    static constexpr int NUM_SAMPLE_POINTS = 5;
+    // Spacing between sample points in metres
+    static constexpr double SAMPLE_POINTS_SPACING_M = 0.1;
+
+    std::optional<Shot> best_shot = std::nullopt;
+
+    // Vector representing the line on which to sample shot origin points
+    Vector sampling_vector =
+        (field.enemyGoalCenter() - starting_point).perpendicular().normalize();
+
+    for (int step = -NUM_SAMPLE_POINTS; step <= NUM_SAMPLE_POINTS; ++step)
+    {
+        Point shot_origin =
+            starting_point + (sampling_vector * step * SAMPLE_POINTS_SPACING_M);
+
+        if (!contains(field.fieldLines(), shot_origin) ||
+            field.pointInEnemyDefenseArea(shot_origin))
+        {
+            continue;
+        }
+
+        std::optional<Shot> shot =
+            calcBestShotOnGoal(field, friendly_team, enemy_team, shot_origin, goal,
+                               robots_to_ignore, radius);
+
+        if (shot && (!best_shot || shot->getOpenAngle() > best_shot->getOpenAngle()))
+        {
+            best_shot = shot;
+        }
+    }
```

Comment:
Suggestion: I can recommend trying performance of using `std::transform` and then calling `std::sort` on the list which would looker cleaner potentially , which should be O(n*log(n)), as the sort dominates - call me out if I'm wrong here on the analysis.

That would give you all the sampled shots ranked, which might be useful information later in the future.

Reply:
now that I think of it, it would be really helpful to accumulate & return a list of all sampled shots that have a direct line that scores, and sort by some "feasibility" depending on the enemy defence positioning.

Reply:
`std::transform` is O(n) and will dominate over O(nlogn), and this for loop is O(n). Time complexity doesn't really matter here because n is always smallish. I'll convert this to using std algorithms anyways for stylistic purposes.

What potential use cases are you thinking of for returning a list of all feasible sampled shots? Ideally we always want to take the most feasible shot, so I'm not sure if there's a practical reason why we would need the 2nd, 3rd, 4th, etc. best shot